modules/caddyhttp/matchers.go

```diff
@@ -1051,7 +1223,356 @@ func (mre *MatchRegexp) UnmarshalCaddyfile(d *caddyfile.Dispenser) error {
+		if !ok {
+			return nil, errors.New("missing 'request' argument")
+		}
+		varNames := nsAttr.CandidateVariableNames()
+		if len(varNames) != 1 || len(varNames) == 1 && varNames[0] != "request" {
+			return nil, errors.New("missing 'request' argument")
+		}
+		matcherData, ok := callArgs[1].(interpreter.InterpretableConst)
+		if !ok {
+			return nil, errors.New("matcher arguments must be constant values")
+		}
+		matcher, err := fac(matcherData)
+		if err != nil {
+			return nil, err
+		}
+		return interpreter.NewCall(
+			i.ID(), funcName, funcName+"_opt",
+			[]interpreter.Interpretable{reqAttr},
+			func(args ...ref.Val) ref.Val {
+				// The request value, guaranteed to be of type celHTTPRequest
+				celReq := args[0]
+				// If needed this call could be changed to convert the value
+				// to a *http.Request using CEL's ConvertToNative method.
+				request := celReq.Value().(celHTTPRequest)
+				return types.Bool(matcher.Match(request.Request))
+			}), nil
+	}
+}
+
+// celMatcherStub is a stubbed out implementation of the function call generated
+// during macro expansion.
+//
+// The implementation will return an error as the decorator is expected to replace
+// this stub with a implementation that precompiles the backing matcher.
+func celMatcherStub(funcName string) functions.BinaryOp {
+	return func(rhs, lhs ref.Val) ref.Val {
+		return types.NewErr("function '%s' not implemented, error in program plan step.", funcName)
+	}
+}
+
+// celMatcherStringListMacroExpander validates that the macro is called
+// with a variable number of string arguments (at least one).
+//
+// The arguments are collected into a single list argument the following
+// function call returned: <funcName>(request, [args])
+func celMatcherStringListMacroExpander(funcName string) parser.MacroExpander {
+	return func(eh parser.ExprHelper, target *exprpb.Expr, args []*exprpb.Expr) (*exprpb.Expr, *common.Error) {
+		matchArgs := []*exprpb.Expr{}
+		if len(args) == 0 {
+			return nil, &common.Error{
+				Message: "matcher requires at least one argument",
+			}
+		}
+		for _, arg := range args {
+			if isCELStringLiteral(arg) {
+				matchArgs = append(matchArgs, arg)
+			} else {
+				return nil, &common.Error{
+					Location: eh.OffsetLocation(arg.GetId()),
+					Message:  "matcher arguments must be string constants",
+				}
+			}
+		}
+		return eh.GlobalCall(funcName, eh.Ident("request"), eh.NewList(matchArgs...)), nil
+	}
+}
+
+// celMatcherStringMacroExpander validates that the macro is called a single
+// string argument.
+//
+// The following function call is returned: <funcName>(request, arg)
+func celMatcherStringMacroExpander(funcName string) parser.MacroExpander {
+	return func(eh parser.ExprHelper, target *exprpb.Expr, args []*exprpb.Expr) (*exprpb.Expr, *common.Error) {
+		for _, arg := range args {
+			if isCELStringLiteral(arg) {
+				return eh.GlobalCall(funcName, eh.Ident("request"), arg), nil
+			}
+			return nil, &common.Error{
```

Comment:
The linter complains in here:

https://github.com/caddyserver/caddy/pull/4715/files#diff-906c27e0b388ea54848a849573bdf556434597d21b03514c4bea5a07989b7206R1414

> modules/caddyhttp/matchers.go#L1414
> SA4004: the surrounding loop is unconditionally terminated (staticcheck)

I figure `len(args) == 1` should be checked instead of doing a loop, and just use `args[0]` afterwards

Reply:
Fixed.